File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) <= 1:
            return 0

        n = len(prices)

        dp = [[0] * n for _ in range(3)]

        for i in range(1, 3):
            maxDiff = -prices[0]
            for j in range(1, n):
                dp[i][j] = max(dp[i][j - 1], prices[j] + maxDiff)

                maxDiff = max(maxDiff, dp[i - 1][j] - prices[j])

        return dp[2][n - 1]

File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length <= 1) {
            return 0;
        }

        int n = prices.length;

        int[][] dp = new int[3][n];

        for (int i = 1; i <= 2; i++) {
            int maxDiff = -prices[0];
            for (int j = 1; j < n; j++) {
                dp[i][j] = Math.max(dp[i][j - 1], prices[j] + maxDiff);
                maxDiff = Math.max(maxDiff, dp[i - 1][j] - prices[j]);
            }
        }

        return dp[2][n - 1];
    }
}